native/kotlin/api/kotlin/src/main/kotlin/rs/wordpress/api/kotlin/WpRequestExecutor.kt

```diff
@@ -91,4 +116,83 @@ class WpRequestExecutor(
     override suspend fun sleep(millis: ULong) {
         delay(millis.toLong())
     }
+
+    private fun getClient(): OkHttpClient {
+        if (allowedHostnames.isEmpty()) {
+            return okHttpClient
+        }
+
+        val clientCertificates = HandshakeCertificates.Builder()
+            .addPlatformTrustedCertificates()
+            .addInsecureHost(allowedHostnames.first())
+            .build()
+
+        return okHttpClient.newBuilder()
+            .hostnameVerifier(WpRequestExecutorHostnameVerifier(allowedHostnames))
+            .sslSocketFactory(clientCertificates.sslSocketFactory(), clientCertificates.trustManager)
+            .build()
+    }
+
+    @Suppress("UnusedParameter")
+    private fun handleInvalidSSLError(e: SSLPeerUnverifiedException, request: Request) {
+        // It's kind of weird and annoying that we need to make a second request to get
+        // this data, but it doesn't seem like we can get it from the response or the
+        // `SSLPeerUnverifiedException` directly.
+        val url = request.url.toUrl()
+
+        // We spin up a new connection that'll accept any certificate. The connection will then
+        // contain all the details we need for the error.
+        val newConnection = url.openConnection() as HttpsURLConnection
+        newConnection.setHostnameVerifier { _, _ -> return@setHostnameVerifier true }
+        newConnection.connect()
+
+        // Certificate is parsed by the Rust shared implementation.
+        val certificates = newConnection.serverCertificates.map { parseCertificate(it.encoded) }
+
+        if (certificates.isEmpty()) {
+            val error = RequestExecutionErrorReason.InvalidSslError(
+                siteCertificate = null,
+                certificateChain = emptyList(),
+                errorMessage = "Invalid certificate for domain",
+                suggestedAction = null
+            )
+
+            throwExceptionFor(error)
+        }
+
+        val siteCertificate = certificates.first()
+
+        val error = RequestExecutionErrorReason.InvalidSslError(
+            siteCertificate = siteCertificate,
+            certificateChain = certificates.mapNotNull { it },
+            errorMessage = "Invalid certificate for domain",
+            suggestedAction = null
+        )
+
+        throwExceptionFor(error)
+    }
+
+    @Suppress("UnusedParameter")
+    private fun handleUnknownHostException(e: UnknownHostException, request: Request) {
+        val error = RequestExecutionErrorReason.NonExistentSiteError(
+            errorMessage = e.localizedMessage,
+            suggestedAction = "Check that the URL is valid and try again"
+        )
+
+        throwExceptionFor(error)
+    }
+
+    private fun throwExceptionFor(reason: RequestExecutionErrorReason) {
```

Comment:
I think it'd be better to implement a `buildRequestExecutionException` instead. It also ties back to my previous suggestion about removing the unnecessary `error("Unknown Error occurred")` line from `execute` function.